File: apps/web/src/it/receiving/receiving.ts
// The Receiving Yard's read model — the INBOUND third of the operator
// surface, upstream of the Marshalling Yard and the Crew Board.
//
// In a rail yard the receiving yard is where inbound trains stand
// before anything is sorted: this page is every packet that ASKS the
// platform for something — feedback, an alarm, a finding a session
// filed, a design question, a protocol step — from the moment it is
// opened until an actor takes it into a build, a review or a decision.
// David, 2026-09-12: "We haven't fixed up visualization upstream of the
// Crew Board that helps us see inbound jobs better." The prototype this
// ports ran on a live read and found eight of his own feedback packets
// standing, the oldest 21 days, and 67 of a week's 267 arrivals naming
// no channel at all.
//
// WHAT COUNTS AS INBOUND is derived from the workflow registry, not
// listed here by kind: a platform-category kind that is neither a chore
// (`maintenance-*`) nor the delivery pipeline. The pipeline set below
// is the one list this file keeps, and it is short because the
// pipeline's kinds are the ones the Train Yard already shows.
//
// A LANE IS WHAT THE SERVER READ, AND ONLY WHAT THE FILER RECORDED.
// Each row carries `lane: {lane, basis}` because this page asks the list
// for it (`lane=true`): the server reads the lane the filer recorded, in
// the one vocabulary (`boss_jobs::channels::lane_of`), and a packet whose
// filer named none is `unclassified` — never guessed into a lane, because
// the whole reason this page exists is to make "where does our work come
// from" a fact instead of a feeling. Until backlog 1eea4554 this file
// kept a six-lane vocabulary and three kind lists of its own and read a
// key no filer writes, so 0 of 1,626 arrivals read as recorded while 492
// of them had recorded one; `regions/receiving.rs` pins that no rule
// grows back here.
//
// NO NUMBER THIS SURFACE MAKES UP. A page that truncated (`total` past
// the rows read) is reported as such by the page; a failed read is a
// failure, never an empty track.

import { isHumanActor } from '../../data/actor';
import { fetchRemote, type Remote } from '../../data/remote';
import { failedVerb, type FailedVerb } from '../../steps/failedVerb';

// ---------------------------------------------------------------------
// Which kinds are inbound
// ---------------------------------------------------------------------

/** The delivery pipeline: a packet of these kinds is a car in transit or
 *  the machinery moving it, never a request. The Train Yard shows them. */
export const PIPELINE_KINDS: ReadonlySet<string> = new Set([
  'pr-train',
  'gate-run',
  'ship-a-change',
  'ops-request',
  'park-a-job',
  'emergency-merge',
  'repair-a-train',
  'publish-request',
  'regenerate-deployment',
]);

export type WorkflowRow = Readonly<{ kind: string; category: string; status: string }>;

const str = (v: unknown): string | null => (typeof v === 'string' && v !== '' ? v : null);

export function parseWorkflows(raw: unknown): ReadonlyArray<WorkflowRow> {
  const list = Array.isArray(raw)
    ? raw
    : Array.isArray((raw as { data?: unknown } | null)?.data)
      ? ((raw as { data: unknown[] }).data)
      : [];
  return list.flatMap((w) => {
    const r = w as Record<string, unknown>;
    const kind = str(r.kind);
    return kind
      ? [{ kind, category: str(r.category) ?? '', status: str(r.status) ?? 'active' }]
      : [];
  });
}

/** Active platform kinds that are neither chores nor the pipeline, sorted. */
export function inboundKinds(workflows: ReadonlyArray<WorkflowRow>): ReadonlyArray<string> {
  return [
    ...new Set(
      workflows
        .filter((w) => w.status === 'active' && w.category === 'platform')
        .map((w) => w.kind)
        .filter((k) => !k.startsWith('maintenance-') && !PIPELINE_KINDS.has(k)),
    ),
  ].sort();
}

// ---------------------------------------------------------------------
// Lane — the server's reading, drawn as it came
// ---------------------------------------------------------------------

export type LaneBasis = 'recorded' | 'unclassified';

export type LaneReading = Readonly<{ lane: string; basis: LaneBasis }>;

/** What the server answers for a packet whose filer recorded no lane. */
export const UNCLASSIFIED: LaneReading = { lane: 'unclassified', basis: 'unclassified' };

/** The `lane` the server put on a listed row. The read asks for it, and a
 *  server that cannot answer `lane=true` refuses the whole read (400),
 *  so a row without one is not something this page's read returns; it
 *  is drawn unclassified rather than dropped, because a lost packet
 *  would shrink every count on the page. */
export function laneOf(row: unknown): LaneReading {
  const l = ((row ?? {}) as { lane?: unknown }).lane as Record<string, unknown> | undefined;
  const lane = str(l?.lane);
  return lane !== null && l?.basis === 'recorded' ? { lane, basis: 'recorded' } : UNCLASSIFIED;
}

/** The lanes to draw: every recorded lane the rows carry, in the
 *  server's spelling and alphabetical, then `unclassified` — always,
 *  because it is the reading this page exists to make, and an empty
 *  track there is an answer. */
export function lanesOf(rows: ReadonlyArray<InboundRow>): ReadonlyArray<string> {
  const recorded = [
    ...new Set(rows.filter((r) => r.laneBasis === 'recorded').map((r) => r.lane)),
  ].sort();
  return [...recorded, UNCLASSIFIED.lane];
}

/** A hue per lane, by the lane's place in [`lanesOf`], so one reading
 *  paints a lane alike in the chart, the tracks and the manifest. Ten,
 *  for the ten lanes a filer may name; `unclassified` is the faint one,
 *  because it is the absence of an answer. */
const LANE_COLORS: ReadonlyArray<string> = [
  'var(--signal)',
  'var(--err)',
  'var(--warn)',
  'var(--ok)',
  'var(--busy)',
  'var(--map-line-publish)',
  'var(--map-line-tenant)',
  'var(--fog)',
  'var(--map-line-delivery)',
  'var(--map-line-siding)',
];

export function laneColor(lane: string, lanes: ReadonlyArray<string>): string {
  const i = lanes.indexOf(lane);
  return lane === UNCLASSIFIED.lane || i < 0
    ? 'var(--text-faint)'
    : (LANE_COLORS[i % LANE_COLORS.length] ?? 'var(--text-faint)');
}

/** The kind a person files through the feedback door. The feedback
 *  reading counts it by KIND, not by lane: measured 2026-09-26, 17 of the
 *  19 user-feedback packets in the window recorded no lane, and reading a
 *  lane off the kind is the rule this file stopped keeping. */
export const FEEDBACK_KIND = 'user-feedback';

// ---------------------------------------------------------------------
// Rows — one parse at the fetch site
// ---------------------------------------------------------------------

/** A ready step, and — when the ops verb it waits on FAILED — the
 *  note the dispatcher left on it (backlog 074e1287): the step is
 *  still ready, the packet is troubled, and the manifest must say so. */
export type ReadyStep = Readonly<{ kind: string; who: string | null; failed: FailedVerb | null }>;

export type InboundRow = Readonly<{
  id: string;
  kind: string;
  title: string;
  status: 'open' | 'closed';
  openedOn: string;
  closedOn: string | null;
  priority: string;
  /** The lane the server read ([`laneOf`]). */
  lane: string;
  laneBasis: LaneBasis;
  ready: ReadonlyArray<ReadyStep>;
  /** An actor has picked it up: a step other than its trigger has
   *  completed ([`takenIn`]). From then on it is marshalling's, not
   *  standing here (design 62de32ae decision 4). */
  takenIn: boolean;
}>;

/** THE INTAKE RULE — `regions.rs::taken_in`, the server's partition
 *  between receiving and marshalling, read the same way here so the
 *  receiving floor stands what the region's header counts. The intake
 *  step is the first step after the trigger, read off the step kinds:
 *  a `triage`, a `decide`, whatever a protocol put first. The trigger
 *  completes at admission, so it takes nothing in; a skipped step is
 *  not a completed one. Measured 2026-09-24: 216 of 232 open
 *  backlog-items were triaged and waiting on `build`, and this board
 *  stood every one of them. */
export const TRIGGER_STEP_KIND = 'trigger';

export function takenIn(steps: ReadonlyArray<Readonly<Record<string, unknown>>>): boolean {
  return steps.some((s) => s.status === 'completed' && s.kind !== TRIGGER_STEP_KIND);
}

export type JobsPage = Readonly<{ rows: ReadonlyArray<InboundRow>; total: number }>;

export function parseJobsPage(raw: unknown): JobsPage {
  const env = raw as { data?: unknown; total?: unknown } | null;
  const data = Array.isArray(env?.data) ? (env.data as ReadonlyArray<unknown>) : [];
  const total = typeof env?.total === 'number' ? env.total : 0;
  const rows = data.flatMap((j): InboundRow[] => {
    const r = j as Record<string, unknown>;
    const id = str(r.id);
    const kind = str(r.kind);
    const openedOn = str(r.opened_on);
    if (!id || !kind || !openedOn) return [];
    const steps = Array.isArray(r.steps) ? (r.steps as ReadonlyArray<Record<string, unknown>>) : [];
    const ln = laneOf(r);
    return [
      {
        id,
        kind,
        title: str(r.title) ?? '',
        status: r.status === 'closed' ? 'closed' : 'open',
        openedOn,
        closedOn: str(r.closed_on),
        priority: str(r.priority) ?? 'standard',
        lane: ln.lane,
        laneBasis: ln.basis,
        ready: steps
          .filter((s) => s.status === 'ready')
          .map((s) => ({
            kind: str(s.kind) ?? '',
            who: str(s.assignee_id),
            failed: failedVerb(s.metadata),
          })),
        takenIn: takenIn(steps),
      },
    ];
  });
  return { rows, total };
}

// ---------------------------------------------------------------------
// Holder, age, band
// ---------------------------------------------------------------------

export type Holder = Readonly<{ who: 'agent' | 'human' | 'nobody'; label: string }>;

/** Who the packet waits on. A person on any ready step outranks the
 *  agent: the packet is standing for THEIR answer. */
export function holderOf(row: InboundRow): Holder {
  if (row.ready.length === 0) return { who: 'nobody', label: 'no ready step' };
  const human = row.ready.find((s) => s.who !== null && isHumanActor(s.who));
  if (human?.who) return { who: 'human', label: human.who };
  const agent = row.ready.find((s) => s.who !== null);
  if (agent) return { who: 'agent', label: 'the agent' };
  return { who: 'nobody', label: 'unassigned' };
}

/** The first ready step whose verb FAILED, or null for a packet with
 *  none: what the manifest's troubled reading and the track's car
 *  title both read. Measured 2026-09-19: publish 254177e2's open-pr
 *  sat annotated for hours and this yard drew a healthy publish. */
export function failedStep(row: InboundRow): ReadyStep | null {
  return row.ready.find((s) => s.failed !== null) ?? null;
}

const DAY_MS = 86_400_000;
const utc = (day: string): number => Date.parse(`${day}T00:00:00Z`);

export function ageDays(openedOn: string, today: string): number {
  return Math.max(0, Math.round((utc(today) - utc(openedOn)) / DAY_MS));
}

/** The Marshalling Yard's proposed thresholds, shown here until the
 *  packet carries its own: triage within 3 days, anything within 14. */
export const AGE_THRESHOLDS = { aging: 3, stale: 14 } as const;

export type AgeBand = 'fresh' | 'aging' | 'stale';

export function ageBand(days: number): AgeBand {
  if (days > AGE_THRESHOLDS.stale) return 'stale';
  if (days > AGE_THRESHOLDS.aging) return 'aging';
  return 'fresh';
}

/** ISO days, oldest first, the last `n` ending on `today`. */
export function daysEndingOn(today: string, n: number): ReadonlyArray<string> {
  return Array.from({ length: n }, (_, i) =>
    new Date(utc(today) - (n - 1 - i) * DAY_MS).toISOString().slice(0, 10),
  );
}

// ---------------------------------------------------------------------
// The two views: arrivals per day, and what is standing
// ---------------------------------------------------------------------

export type DayFlow = Readonly<{
  day: string;
  arrived: number;
  byLane: Readonly<Record<string, number>>;
  /** Closed that day — including packets that arrived before the window. */
  left: number;
}>;

export function arrivalsByDay(
  rows: ReadonlyArray<InboundRow>,
  days: ReadonlyArray<string>,
): ReadonlyArray<DayFlow> {
  return days.map((day) => {
    const arrived = rows.filter((r) => r.openedOn === day);
    const byLane = arrived.reduce<Record<string, number>>(
      (acc, r) => ({ ...acc, [r.lane]: (acc[r.lane] ?? 0) + 1 }),
      {},
    );
    return {
      day,
      arrived: arrived.length,
      byLane,
      left: rows.filter((r) => r.closedOn === day).length,
    };
  });
}

export type WaitingRow = InboundRow &
  Readonly<{ age: number; band: AgeBand; holder: Holder }>;

/** A packet standing in receiving: open, and not yet taken in. */
export const standsHere = (r: InboundRow): boolean => r.status === 'open' && !r.takenIn;

/** Every packet standing in receiving, oldest first. */
export function waiting(rows: ReadonlyArray<InboundRow>, today: string): ReadonlyArray<WaitingRow> {
  return rows
    .filter(standsHere)
    .map((r) => {
      const age = ageDays(r.openedOn, today);
      return { ...r, age, band: ageBand(age), holder: holderOf(r) };
    })
    .sort((a, b) => b.age - a.age || a.kind.localeCompare(b.kind));
}

// ---------------------------------------------------------------------
// What the snapshot says
// ---------------------------------------------------------------------

export type Readings = Readonly<{
  feedbackStanding: Readonly<{ count: number; oldestDays: number }>;
  onOneActor: Readonly<{ count: number; of: number; actor: string }>;
  unrecorded: Readonly<{ count: number; of: number }>;
}>;

export function readings(
  rows: ReadonlyArray<InboundRow>,
  today: string,
  days: ReadonlyArray<string>,
): Readings {
  const standing = waiting(rows, today);
  const feedback = standing.filter((r) => r.kind === FEEDBACK_KIND);
  const holders = standing.map((r) => r.holder.label);
  const top = [...new Set(holders)]
    .map((label) => ({ label, n: holders.filter((h) => h === label).length }))
    .sort((a, b) => b.n - a.n)[0];
  const window = new Set(days);
  const arrived = rows.filter((r) => window.has(r.openedOn));
  return {
    feedbackStanding: {
      count: feedback.length,
      oldestDays: feedback.reduce((m, r) => Math.max(m, r.age), 0),
    },
    onOneActor: { count: top?.n ?? 0, of: standing.length, actor: top?.label ?? '—' },
    unrecorded: {
      count: arrived.filter((r) => r.laneBasis === 'unclassified').length,
      of: arrived.length,
    },
  };
}

// ---------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------

/** The registry: which kinds exist and what category each is. */
export function loadWorkflows(): Promise<
  Exclude<Remote<ReadonlyArray<WorkflowRow>>, { kind: 'loading' }>
> {
  return fetchRemote('/api/workflows', parseWorkflows);
}

/** One kind's packets: everything open, plus what closed inside the
 *  window. Real work only — the demo tenant's packets are not inbound.
 *  `lane=true` asks the server to put its lane reading on each row. */
export function loadKind(
  kind: string,
  windowDays: number,
  limit: number,
  offset = 0,
): Promise<Exclude<Remote<JobsPage>, { kind: 'loading' }>> {
  return fetchRemote(
    `/api/jobs?kind=${encodeURIComponent(kind)}&simulated=false&closed_within=${windowDays}&lane=true&limit=${limit}&offset=${offset}`,
    parseJobsPage,
  );
}

/** EVERY packet of one kind in the window, page after page until the
 *  server's `total` — a limit is not a filter (design 62de32ae
 *  decision 4). This board read one page of 500 and, measured
 *  2026-09-24, backlog-item had 822 in the window: every count below
 *  it was a floor, said in a notice nobody reads as a number. A row a
 *  shifting page repeats is kept once. Any page failing fails the
 *  whole read — half a kind is not a kind. */
export async function loadEveryPage(
  kind: string,
  windowDays: number,
  pageSize: number,
  load: typeof loadKind = loadKind,
): Promise<Exclude<Remote<JobsPage>, { kind: 'loading' }>> {
  const read = async (
    offset: number,
    got: ReadonlyArray<InboundRow>,
  ): Promise<Exclude<Remote<JobsPage>, { kind: 'loading' }>> => {
    const page = await load(kind, windowDays, pageSize, offset);
    if (page.kind === 'failed') return page;
    const seen = new Set(got.map((r) => r.id));
    const rows = [...got, ...page.data.rows.filter((r) => !seen.has(r.id))];
    const next = offset + page.data.rows.length;
    return page.data.rows.length === 0 || next >= page.data.total
      ? { kind: 'ready', data: { rows, total: page.data.total } }
      : read(next, rows);
  };
  return read(0, []);
}

File: crates/core/boss-jobs/src/regions/receiving.rs
//! RECEIVING: inbound packets standing until an actor takes them in —
//! with the receiving yard's pipeline set and age bands, the second of
//! the two client-side predicates ported here once (`receiving.ts`) and
//! pinned against the TypeScript.

use super::*;
use crate::channels::{LaneBasis, lane_of};

/// The delivery pipeline's kinds — a packet of these is a car in
/// transit or the machinery moving it, never a request, so the
/// receiving yard leaves them to the train yard. `receiving.ts::
/// PIPELINE_KINDS`, ported; the test `the_pipeline_kinds_match_the_
/// receiving_yard` holds the two equal.
pub const PIPELINE_KINDS: [&str; 9] = [
    "pr-train",
    "gate-run",
    "ship-a-change",
    "ops-request",
    "park-a-job",
    "emergency-merge",
    "repair-a-train",
    "publish-request",
    "regenerate-deployment",
];

/// The receiving yard's age bands, in days from `opened_on`: triage
/// within 3 days, anything within 14 (`receiving.ts::AGE_THRESHOLDS`,
/// ported and pinned by the same test).
pub const AGING_DAYS: i64 = 3;
pub const STALE_DAYS: i64 = 14;

/// Which kinds are inbound: active platform-category kinds that are
/// neither chores (`maintenance-*`) nor the pipeline
/// (`receiving.ts::inboundKinds`). Derived from the registry, never
/// listed by kind.
pub fn inbound_kinds(workflows: &[WorkflowSpec]) -> Vec<String> {
    let mut kinds: Vec<String> = workflows
        .iter()
        .filter(|w| w.status == crate::registry::WorkflowStatus::Active)
        .filter(|w| w.category == "platform")
        .filter(|w| {
            !w.kind.starts_with("maintenance-") && !PIPELINE_KINDS.contains(&w.kind.as_str())
        })
        .map(|w| w.kind.clone())
        .collect();
    kinds.sort();
    kinds.dedup();
    kinds
}

/// RECEIVING: inbound packets standing — feedback, alarms, findings,
/// design questions — from the moment they open until an actor takes
/// them in: until the intake step completes ([`taken_in`]), after which
/// the packet is marshalling's, never both ([`members`]). The receiving
/// yard's own bands decide the state: any
/// packet older than [`STALE_DAYS`] is troubled, older than
/// [`AGING_DAYS`] attention. The trend is inbound arrivals per day.
pub(super) fn receiving(inputs: &RegionInputs<'_>, w: &Windows) -> Region {
    const UNIT: &str = "packets standing";
    let Some(inbound) = inputs.inbound else {
        return region(
            "receiving",
            None,
            None,
            UNIT,
            unread_settled(
                "the workflow registry that names the inbound kinds could not be read",
                inputs.now,
            ),
            Trend {
                metric: "inbound".to_string(),
                unit: "per day".to_string(),
                current: None,
                previous: None,
                samples: 0,
                previous_samples: 0,
            },
            vec![measure("oldest untriaged", None, "days")],
        );
    };
    let (cur, prev) = count_split(w, inbound.iter().filter_map(|(j, _)| opened_at(j)));
    let trend = rate_trend("inbound", w, cur, prev);
    let today = inputs.now.date_naive();
    // THE PARTITION (design 62de32ae decision 4): standing here means not
    // yet taken in, and held by no region of its own.
    let open: Vec<&Job> = receiving_standing(inputs).unwrap_or_default();
    let oldest_day = open.iter().map(|j| j.opened_on).min();
    let oldest = oldest_day.map_or(0, |d| (today - d).num_days());
    let n = open.len();
    // A band in whole days from `opened_on` is crossed at the midnight
    // that makes the age exceed it — which IS its onset, read off the
    // packet's own date.
    let crossed = |band_days: i64| {
        oldest_day
            .and_then(|d| d.checked_add_days(chrono::Days::new(u64::try_from(band_days + 1).ok()?)))
            .and_then(|d| d.and_hms_opt(0, 0, 0))
            .map(|t| chrono::DateTime::from_naive_utc_and_offset(t, chrono::Utc))
    };
    let mut findings = Vec::new();
    if oldest > STALE_DAYS {
        findings.push(Finding::new(
            bands::RECEIVING_STALE,
            crossed(STALE_DAYS),
            format!("oldest {oldest}d"),
            format!(
                "{} standing, the oldest {oldest} days (past the {STALE_DAYS}-day band)",
                plural(n, "packet", "packets")
            ),
        ));
    }
    if oldest > AGING_DAYS {
        findings.push(Finding::new(
            bands::RECEIVING_AGING,
            crossed(AGING_DAYS),
            format!("oldest {oldest}d"),
            format!(
                "{} standing, the oldest {oldest} days (past the {AGING_DAYS}-day triage band)",
                plural(n, "packet", "packets")
            ),
        ));
    }
    let settled = settle(
        findings,
        format!("{} standing", plural(n, "packet", "packets")),
        inputs.now,
    );
    // THE KPI (decision 9): the oldest untriaged, and the share of what
    // stands whose filer recorded no lane.
    #[allow(clippy::cast_precision_loss)]
    let kpi = vec![
        match oldest_day {
            Some(_) => measure("oldest untriaged", Some(oldest as f64), "days"),
            None => measure_said(
                "oldest untriaged",
                None,
                "days",
                "nothing untriaged".to_string(),
            ),
        },
        unrecorded_share(&open),
    ];
    region("receiving", Some(n), None, UNIT, settled, trend, kpi)
}

/// THE SHARE WHOSE CHANNEL IS UNRECORDED — decision 9's second half,
/// named residue on backlog c3105b2a until the lane rule moved to the
/// server (backlog 1eea4554): it was the client's (`receiving.ts::
/// channelOf`), in a six-lane vocabulary of its own, so porting it would
/// have been a second copy to pin. Now both read [`lane_of`], the one
/// rule, and the board draws the lane the jobs list classified.
/// Measured over what STANDS, the region's own count, as a whole percent.
/// Nothing standing is no reading, never a 0% that says every filer
/// named their lane.
fn unrecorded_share(standing: &[&Job]) -> Measure {
    const NAME: &str = "channel unrecorded";
    let of = standing.len();
    if of == 0 {
        return measure_said(NAME, None, "%", format!("{NAME}: nothing standing"));
    }
    let unrecorded = standing
        .iter()
        .filter(|j| lane_of(&j.metadata).basis == LaneBasis::Unclassified)
        .count();
    #[allow(clippy::cast_precision_loss)]
    let pct = (unrecorded as f64 * 100.0 / of as f64).round();
    measure_said(
        NAME,
        Some(pct),
        "%",
        format!(
            "{NAME} {}% ({unrecorded} of {of} standing)",
            number_text(pct)
        ),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::regions::fixtures::*;

    /// Receiving: the age bands from `receiving.ts`, and arrivals per
    /// day from the packets' own open stamps.
    #[test]
    fn receiving_reads_the_age_bands_and_inbound_arrivals() {
        let mut fresh = job(
            "user-feedback",
            "fresh",
            JobStatus::Open,
            json!({ "opened_at": "2026-09-19T09:00:00Z" }),
        );
        fresh.opened_on = chrono::NaiveDate::from_ymd_opt(2026, 9, 19).unwrap();
        let mut aging = job("backlog-item", "aging", JobStatus::Open, json!({}));
        aging.opened_on = chrono::NaiveDate::from_ymd_opt(2026, 9, 14).unwrap();
        let mut stale = job("backlog-item", "stale", JobStatus::Open, json!({}));
        stale.opened_on = chrono::NaiveDate::from_ymd_opt(2026, 8, 20).unwrap();
        let mut yesterday = job(
            "user-feedback",
            "done",
            JobStatus::Closed,
            json!({ "opened_at": "2026-09-18T09:00:00Z" }),
        );
        yesterday.opened_on = chrono::NaiveDate::from_ymd_opt(2026, 9, 18).unwrap();
        let status = empty_status();
        let inbound = untaken(vec![fresh.clone(), aging.clone()]);
        let out = regions(&inputs(
            &status,
            &[],
            &[],
            &[],
            &[],
            Some(&inbound),
            Some(&[]),
        ));
        let r = by_name(&out, "receiving");
        assert_eq!(r.count, Some(2));
        // Past the triage band: ATTENTION, with the band named against
        // the number beside it (design 62de32ae, decision 1) — and the
        // band was crossed the midnight the oldest turned four days old.
        assert_eq!(r.state, RegionState::Attention, "{}", r.why);
        let band = r.band.as_ref().unwrap();
        assert_eq!(band.id, "receiving-aging");
        assert_eq!(band.reads, "oldest 5d > the 3-day triage band");
        assert_eq!(band.since.as_deref(), Some("2026-09-18T00:00:00+00:00"));
        assert_eq!(band.held_minutes, Some(36 * 60));
        assert_eq!(r.unit, "packets standing");
        assert_eq!(r.kpi[0].text, "oldest untriaged 5 days");
        assert_eq!(r.trend.current, Some(1.0));
        assert_eq!(r.trend.previous, Some(0.0));

        let inbound = untaken(vec![fresh, aging, stale, yesterday]);
        let out = regions(&inputs(
            &status,
            &[],
            &[],
            &[],
            &[],
            Some(&inbound),
            Some(&[]),
        ));
        let r = by_name(&out, "receiving");
        assert_eq!(r.count, Some(3));
        assert_eq!(r.state, RegionState::Troubled, "{}", r.why);
        assert!(r.why.contains("30 days"), "{}", r.why);
        assert_eq!(r.trend.previous, Some(1.0));
    }

    /// THE KPI's SECOND HALF (design 62de32ae decision 9, ported by
    /// backlog 1eea4554): of the packets standing, the share whose filer
    /// recorded no lane — read through `channels::lane_of`, the one rule
    /// the board's rows are classified by too, so the region and the
    /// board cannot count two different things.
    #[test]
    fn receiving_measures_the_share_of_standing_packets_with_no_recorded_lane() {
        let recorded = job(
            "backlog-item",
            "said",
            JobStatus::Open,
            json!({ "input_channel": "review-finding" }),
        );
        // A key no filer writes, and a near miss: neither is a lane.
        let old_key = job(
            "backlog-item",
            "old key",
            JobStatus::Open,
            json!({ "channel": "monitoring" }),
        );
        let near_miss = job(
            "user-feedback",
            "near miss",
            JobStatus::Open,
            json!({ "input_channel": "telemetry" }),
        );
        let silent = job("design-doc", "silent", JobStatus::Open, json!({}));
        let status = empty_status();
        let inbound = untaken(vec![recorded, old_key, near_miss, silent]);
        let out = regions(&inputs(
            &status,
            &[],
            &[],
            &[],
            &[],
            Some(&inbound),
            Some(&[]),
        ));
        let r = by_name(&out, "receiving");
        assert_eq!(r.kpi.len(), 2, "{:?}", r.kpi);
        let share = &r.kpi[1];
        assert_eq!(share.name, "channel unrecorded");
        assert_eq!(share.unit, "%");
        assert_eq!(share.value, Some(75.0));
        assert_eq!(share.text, "channel unrecorded 75% (3 of 4 standing)");

        // Nothing standing: no share, said so — never a 0% that reads
        // as every filer saying where their work came from.
        let out = regions(&inputs(
            &status,
            &[],
            &[],
            &[],
            &[],
            Some(&untaken(vec![])),
            Some(&[]),
        ));
        let share = &by_name(&out, "receiving").kpi[1];
        assert_eq!(share.value, None);
        assert_eq!(share.text, "channel unrecorded: nothing standing");
    }

    /// The two facts ported from the receiving yard's TypeScript live
    /// twice; this holds them equal (CLAUDE.md §9a). Reads the source
    /// the client is built from, so a kind added to one side and not
    /// the other names itself here.
    #[test]
    fn the_pipeline_kinds_and_age_bands_match_the_receiving_yard() {
        let src = std::fs::read_to_string(
            boss_testing::repo_root().join("apps/web/src/it/receiving/receiving.ts"),
        )
        .expect("receiving.ts is in the tree");
        let set = src
            .split("PIPELINE_KINDS: ReadonlySet<string> = new Set([")
            .nth(1)
            .and_then(|rest| rest.split("]);").next())
            .expect("receiving.ts declares PIPELINE_KINDS as a Set literal");
        let ts_kinds: Vec<&str> = set
            .split(',')
            .map(|s| s.trim().trim_matches('\''))
            .filter(|s| !s.is_empty())
            .collect();
        assert_eq!(
            ts_kinds, PIPELINE_KINDS,
            "receiving.ts::PIPELINE_KINDS drifted from regions.rs::PIPELINE_KINDS"
        );
        assert!(
            src.contains(&format!(
                "AGE_THRESHOLDS = {{ aging: {AGING_DAYS}, stale: {STALE_DAYS} }}"
            )),
            "receiving.ts::AGE_THRESHOLDS drifted from regions.rs::AGING_DAYS / STALE_DAYS"
        );
        // The train-gate keys the yard's `readTrainGate` reads.
        let yard =
            std::fs::read_to_string(boss_testing::repo_root().join("apps/web/src/it/yard/yard.ts"))
                .expect("yard.ts is in the tree");
        for key in [TRAIN_GATE_WAIT_REASON, TRAIN_GATE_FALLBACK] {
            assert!(
                yard.contains(&format!("md.{key}")),
                "yard.ts no longer reads {key}"
            );
        }
    }

    /// THE BOARD DRAWS THE SERVER'S LANE (backlog 1eea4554): the rule is
    /// [`lane_of`] alone, so the receiving yard's TypeScript asks the
    /// list for it (`lane=true`) and keeps no rule of its own — no read
    /// of the recorded key, none of the key no filer writes, no kind
    /// lists. A client that grows its own classification again names
    /// itself here, the way the pipeline kinds above do (CLAUDE.md §9a).
    #[test]
    fn the_receiving_yard_draws_the_lane_the_server_read() {
        let src = std::fs::read_to_string(
            boss_testing::repo_root().join("apps/web/src/it/receiving/receiving.ts"),
        )
        .expect("receiving.ts is in the tree");
        assert!(
            src.contains("&lane=true"),
            "receiving.ts must ask the jobs list for the server's lane"
        );
        for (word, why) in [
            (
                crate::channels::RECORDED_KEY,
                "reads the recorded key itself",
            ),
            ("md.channel", "reads a key no filer writes"),
            ("DESIGN_KINDS", "keeps a kind list"),
            ("PROTOCOL_KINDS", "keeps a kind list"),
            ("MONITORING_KINDS", "keeps a kind list"),
        ] {
            assert!(
                !src.contains(word),
                "receiving.ts {why} ({word}); the lane is the server's (channels::lane_of)"
            );
        }
        // The two bases a reading can have, spelled as the server spells them.
        for basis in [LaneBasis::Recorded, LaneBasis::Unclassified] {
            let word = serde_json::to_value(basis).unwrap();
            let word = word.as_str().unwrap();
            assert!(
                src.contains(&format!("'{word}'")),
                "receiving.ts no longer parses the basis '{word}'"
            );
        }
    }

    #[test]
    fn inbound_kinds_are_active_platform_kinds_minus_chores_and_the_pipeline() {
        let spec = |kind: &str, category: &str| WorkflowSpec {
            kind: kind.into(),
            version: 1,
            status: crate::registry::WorkflowStatus::Active,
            label: kind.into(),
            description: None,
            category: category.into(),
            subject_kinds: vec![],
            steps: vec![],
            metadata_schema: json!({}),
            entitlements: json!({}),
            metadata: json!({}),
            on_complete_create: vec![],
            owning_team: "it".into(),
            authoring_job_id: None,
            created_at: t(NOW),
        };
        let specs = vec![
            spec("user-feedback", "platform"),
            spec("backlog-item", "platform"),
            spec("maintenance-nightly", "platform"),
            spec("pr-train", "platform"),
            spec("sale", "commerce"),
        ];
        assert_eq!(inbound_kinds(&specs), vec!["backlog-item", "user-feedback"]);
    }
}
